merkle/merkle_tree.go

```diff
@@ -0,0 +1,195 @@
+/*
+Ported from https://github.com/gohornet/hornet codebase.
+Original authors: muXxer <mux3r@web.de>
+                  Alexander Sporn <github@alexsporn.de>
+                  Thoralf-M <46689931+Thoralf-M@users.noreply.github.com>
+
+
+Package merkle provides functions and types to deal with the creation and storage of
+Merkle trees, using the secure SHAKE256 KDF implemented in the signing/key package:
+thus not being affected by the the infamous M-Bug.
+*/
+package merkle
+
+import (
+	"math"
+	"runtime"
+	"sync"
+
+	"github.com/iotaledger/iota.go/consts"
+	"github.com/iotaledger/iota.go/kerl"
+	"github.com/iotaledger/iota.go/signing"
+	"github.com/iotaledger/iota.go/signing/key"
+	"github.com/iotaledger/iota.go/trinary"
+)
+
+// MerkleTree contains the Merkle tree used for the coordinator signatures.
+type MerkleTree struct {
+	// The depth of the Merkle tree.
+	Depth int
+	// The root address of the Merkle tree.
+	Root trinary.Hash
+	// Merkle tree layers indexed by their level.
+	Layers map[int]*MerkleTreeLayer
+}
+
+// MerkleTreeLayer contains the nodes of a layer of a Merkle tree.
+type MerkleTreeLayer struct {
+	// The level of the layer in the tree.
+	Level int
+	// The nodes of the layer.
+	Hashes []trinary.Hash
+}
+
+// MilestoneIndex represents an index of a milestone.
+type MilestoneIndex uint32
+
+// calculateAllAddresses calculates all addresses that are used for the Merkle tree of the coordinator.
+func calculateAllAddresses(seed trinary.Hash, securityLvl int, count int) []trinary.Hash {
+	resultLock := &sync.Mutex{}
+	result := make([]trinary.Hash, count)
+
+	wg := sync.WaitGroup{}
+	wg.Add(runtime.NumCPU())
+
+	// calculate all addresses in parallel.
+	input := make(chan MilestoneIndex)
+	for i := 0; i < runtime.NumCPU(); i++ {
+		go func() {
+			defer wg.Done()
+
+			for index := range input {
+				address, err := computeAddress(seed, index, securityLvl)
+				if err != nil {
+					panic(err)
+				}
+				resultLock.Lock()
+				result[int(index)] = address
+				resultLock.Unlock()
+			}
+		}()
+	}
+
+	for index := 0; index < count; index++ {
+		input <- MilestoneIndex(index)
```

Comment:
You removed my code to calculate the progress and especially the remaining time.
That will be a lot of fun for the user to wait several hours without feedback. :)

Can you please add a consumer function for the calculated progress? So the user can decide if he wants to print it or not.

Reply:
Yeah, removed your code as it was turned into a lib function.
I've addressed your comment by turning`CreateMerkleTree` into a variadic function that may receive a slice of progressCallbacks (only the first one will be used though).